src/main/java/com/google/api/codegen/util/NameFormatterMixin.java

```diff
@@ -0,0 +1,73 @@
+/* Copyright 2016 Google Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.api.codegen.util;
+
+/**
+ * NameFormatterMixin is an abstract class that implements NameFormatter and
+ * simply forwards calls to another NameFormatter. This allows a child class
+ * to provide the interface of NameFormatter along with additional functionality.
+ *
+ * Note to future maintainers: This class should only contain methods which
+ * forward on to NameFormatter and nothing else; otherwise, it is no longer
+ * functioning in spirit as a mix-in.
+ */
+public abstract class NameFormatterMixin implements NameFormatter {
```

Comment:
A design pattern is defined in terms of a solution rather than a specific problem that it's trying to solve. This looks like a Decorator, unless you see some subtle reason why this _design_ doesn't match that pattern. Nevertheless, I understand that you want to convey the _intent_ in your class name.

I'm coming around to understanding why you designed this the way you did. It seems like you really want mixin capabilities to combine `SurfaceNamer` and `NameFormatter` functionality together in each concrete `<Language>SurfaceNamer` class, but Java7 forces you to do it in this indirect way. The problem is that it just doesn't look like a mixin and it can't really be used like one. Once we are allowed to use Java8, then I think we can properly realize your vision here.

I'm okay with Delegate, Delegator, Forwarder, or something like that as a compromise.